src/main/kotlin/uk/gov/dwp/dataworks/logging/LoggerLayoutAppender.kt

```diff
@@ -0,0 +1,35 @@
+package uk.gov.dwp.dataworks.logging
+
+import ch.qos.logback.classic.spi.ILoggingEvent
+import ch.qos.logback.core.LayoutBase
+
+/**
+ * Implementation of [LayoutBase] which allows us to parse events and reformat into JSON structured messages.
+ */
+object LoggerLayoutAppender : LayoutBase<ILoggingEvent>() {
+
+    private var start_time_milliseconds = System.currentTimeMillis()
+
+    private fun getDurationInMilliseconds(epochTime: Long): String {
+        val elapsedMilliseconds = epochTime - start_time_milliseconds
+        return elapsedMilliseconds.toString()
+    }
+
+    override fun doLayout(event: ILoggingEvent?): String {
+        if (event == null) {
+            return ""
+        }
+        return """
+            |{
+                |"timestamp":"${epochToUTCString(event.timeStamp)}",
+                |"log_level":"${event.level}",
+                |"message":${flattenString(event.formattedMessage)},
+                |"exception":"${throwableProxyEventToString(event)}",
+                |"thread":"${event.threadName}",
+                |"logger":"${event.loggerName}",
+                |"duration_in_milliseconds":"${getDurationInMilliseconds(event.timeStamp)}",
+                |${LogFields.asJson}
+            |}
+        """.trimMargin("|").replace("\n", "")
+    }

```

Comment:
Trim margin is very slow, can we have a string builder please? there should be no need for a trim or newline replacing.

Reply:
I see your point. I wanted to strike a balance between code readability and performance. Note that the performance impact is minor in this case and [string templates compile to builders](https://www.baeldung.com/kotlin-concatenate-strings#using-string-templates).
I will remove the post-formatting from the template and add to the KDoc the fields which are appended. Hopefully this will satisfy both our wants.

Reply:
Spoke to Mark and we are happy with this approach